### ISSUE TITLE
Handle messages asynchronously

### ISSUE THREAD

**Issue by ivanovaleksey** [eclipse/paho.mqtt.rust#5]
Hello,
thank you for the great library!

I am looking for Rust MQTT library for my application. 
I want it to be able to handle incoming messages asynchronously because there will be DB interaction and other blocking things.

I played a little with `async_subscribe`, in particular I added `thread::sleep` within `set_message_callback` callback.
And message handling had been queued: one message was handling (i.e. waiting on `sleep`) and new incoming message was queued and processed only after awaking.

Is there any way to have asynchronous message handling?
Should I create some kind of a thread poll within `set_message_callback` and just push incoming message to that pool?
Or it will affect dealing with QoS?
As far as I understand for `QoS=1` we firstly send PUBACK and then call user's callback.
Does it mean that it doesn't matter whether I would handle incoming message in the same or in separate thread because PUBACK was already sent?

Thank you.

**Comment by fpagliughi** [comment]
Yes, you're thinking the right thing.

The message callback happens in the context of a thread in the underlying C library, and it is best not to block it too long. The callback is given ownership of the Message, so you can do with it what you want. The easy thing is just put it into a mpsc::channel, and start another thread (or pool) to process them. Something like:
```
let (tx, rx): (Sender<Option<Message>>, Receiver<Option<Message>>) = mpsc::channel();

client.set_message_callback(move |_,msg| {
    tx.send(msg).unwrap();
});
```

This is actually in the library as a "consuming" client. It is more normally used by the synchronous client, but can be used by the async one as well. Use the `start_consuming()` function. It will install a callback to queue the messages (as shown above) and return the receiving end of the channel which can then be given to a thread to process.

Look at the `sync_consume.rs` example for a little more info.

**Comment by ArthurFleischman** [comment]
> Hello, thank you for the great library!
> 
> I am looking for Rust MQTT library for my application. I want it to be able to handle incoming messages asynchronously because there will be DB interaction and other blocking things.
> 
> I played a little with `async_subscribe`, in particular I added `thread::sleep` within `set_message_callback` callback. And message handling had been queued: one message was handling (i.e. waiting on `sleep`) and new incoming message was queued and processed only after awaking.
> 
> Is there any way to have asynchronous message handling? Should I create some kind of a thread poll within `set_message_callback` and just push incoming message to that pool? Or it will affect dealing with QoS? As far as I understand for `QoS=1` we firstly send PUBACK and then call user's callback. Does it mean that it doesn't matter whether I would handle incoming message in the same or in separate thread because PUBACK was already sent?
> 
> Thank you.

mine message callback doesn't work, how did u managed it?

**Comment by fpagliughi** [comment]
What about it doesn’t work?

**Comment by ArthurFleischman** [comment]
> What about it doesn’t work?

- on_connected
_ on_connection_loss
- on_message

Only the ones attached to the client.start_with_callbacks() works

**Comment by fpagliughi** [comment]
I don't know what's happening here in the wider application:
```
pub fn spawn_mqtt_client(window: Window, client_id: i32, url: String){
    let mut client = &mut MqttClient::new(url.clone(), client_id, window);
    client.setup_client();
    client.init_client();
}
```
You're creating a mutable reference to the client that goes out of scope when the function exits? Where does your client object actually live?

**Comment by fpagliughi** [comment]
I don't know tauri, and just wanted to boil your code down a bit, so I took it out (all 'win' and UserData stuff), and made a single test app with your code. 

```
use paho_mqtt::{self as mqtt, AsyncClient, Message, MQTT_VERSION_5, QOS_1};
use std::time::Duration;
use std::{process, thread};

const QOS: &[i32] = &[1, 1, 1, 1];
pub fn get_topics(id: String) -> [String; 4] {
    println!("gateway-id {}", id);
    [
        format!("status/gateway/{id}"),
        "status/device/+".to_string(),
        "data/device/+".to_string(),
        "test".to_string(),
    ]
}

pub fn create_client(host: String, id: String) -> AsyncClient {
    let create_opts = mqtt::CreateOptionsBuilder::new()
        .client_id(id)
        //.mqtt_version(MQTT_VERSION_5)  <- Not needed anymore. new() works with v3 or v5
        .server_uri(host)
        .finalize();

    // Create the client connection
    AsyncClient::new(create_opts).unwrap_or_else(|e| {
        println!("Error creating the client: {:?}", e);
        process::exit(1);
    })
}

pub fn on_connect_success(cli: &AsyncClient, _msgid: u16) {
    println!("connection success");

    println!("subscribing");
    cli.subscribe_many(&get_topics(cli.client_id()), QOS);
    cli.subscribe("test", 1);

    send_status_msg(cli.client_id(), "online");
}

pub fn on_connect_failure(cli: &AsyncClient, _msgid: u16, rc: i32) {
    println!("Connection attempt failed with error code {}.\n", rc);
    send_status_msg(cli.client_id(), "online");
    thread::sleep(Duration::from_millis(2500));
    cli.reconnect_with_callbacks(on_connect_success, on_connect_failure);
}

pub fn on_connection_loss(cli: &AsyncClient) {
    println!("connenction lost");
    send_status_msg(cli.client_id(), "offline");
    cli.reconnect_with_callbacks(on_connect_success, on_connect_failure);
}

pub fn on_connected(cli: &AsyncClient) {
    send_status_msg(cli.client_id(), "online");
    cli.reconnect_with_callbacks(on_connect_success, on_connect_failure);
}

pub fn on_message(_cli: &AsyncClient, msg_opt: Option<Message>) {
    println!("received msg");
    if let Some(msg) = msg_opt {
        send_data_msg(msg.topic(), msg.payload_str().to_string())
    }
}

fn send_status_msg(id: String, status: &str) {
    println!("[{}] status/gateway/{}", id, status);
}

fn send_data_msg(topic: &str, data: String) {
    println!("{}: {}", topic, data);
}

pub struct MqttClient {
    pub url: String,
    pub client_id: i32,
    client: AsyncClient,
}

impl MqttClient {
    pub fn new(url: String, client_id: i32) -> MqttClient {
        MqttClient {
            url: url.clone(),
            client_id,
            client: create_client(url.clone(), client_id.to_string()),
        }
    }

    pub fn setup_client(&mut self) {
        println!("Setting up client");
        self.client.set_connected_callback(on_connected);
        self.client.set_connection_lost_callback(on_connection_loss);
        self.client.set_message_callback(on_message);
    }

    pub fn init_client(&mut self) {
        let lwt = mqtt::Message::new(
            "test/lwt",
            "[LWT] Gateway subscriber lost connection",
            QOS_1,
        );
        let conn_opts = mqtt::ConnectOptionsBuilder::with_mqtt_version(MQTT_VERSION_5)
            .clean_start(false)
            .properties(mqtt::properties![mqtt::PropertyCode::SessionExpiryInterval => 3600])
            .will_message(lwt)
            .finalize();
        self.client
            .connect_with_callbacks(conn_opts, on_connect_success, on_connect_failure);
    }
}

fn main() {
    let client_id: i32 = 42;
    let url = "mqtt://mqtt.eclipseprojects.io:1883".to_string();

    let mut client = MqttClient::new(url.clone(), client_id);
    client.setup_client();
    client.init_client();

    // Just block to let the
    loop {
        thread::sleep(Duration::from_secs(1))
    }
}
```

When run, this produced:

```
$ cargo run
   Compiling paho_issue_5_reopened v0.1.0 (/home/fmp/tmp/paho_issue_5_reopened)
    Finished dev [unoptimized + debuginfo] target(s) in 0.32s
     Running `target/debug/paho_issue_5_reopened`
Setting up client
connection success
subscribing
gateway-id 42
[42] status/gateway/online
[42] status/gateway/online
Connection attempt failed with error code -3.

[42] status/gateway/online
received msg
connenction lost                                   <- Turned off the network (WiFI)
[42] status/gateway/offline
Connection attempt failed with error code -1.

[42] status/gateway/online
Connection attempt failed with error code -1.

[42] status/gateway/online
Connection attempt failed with error code -1.

[42] status/gateway/online
Connection attempt failed with error code -1.

[42] status/gateway/online                      <- Turned the WiFi back on
connection success
subscribing
gateway-id 42
[42] status/gateway/online
[42] status/gateway/online
...
```

**Comment by fpagliughi** [comment]
Note that it may take a long time for the library/protocol to detect the network loss and fire the "connection lost" callback. It could be 1.5x to 2x the keep alive interval. You don't specify one, and the default is 45 or 60 sec. So it could be up to 2min to fire the callback.

**Comment by ArthurFleischman** [comment]
> Note that it may take a long time for the library/protocol to detect the network loss and fire the "connection lost" callback. It could be 1.5x to 2x the keep alive interval. You don't specify one, and the default is 45 or 60 sec. So it could be up to 2min to fire the callback.

main I committed the slightest of the mistakes, for real, the URL comes form the sqlite database.... i forgot to add the `mqtt://` as the prefix of the `localhost:1883`....  hahahahhahahahahahahahahhahahahah

**Comment by ArthurFleischman** [comment]
> Note that it may take a long time for the library/protocol to detect the network loss and fire the "connection lost" callback. It could be 1.5x to 2x the keep alive interval. You don't specify one, and the default is 45 or 60 sec. So it could be up to 2min to fire the callback.

i think I'm killing the thread, ill check it

**Comment by ArthurFleischman** [comment]
> thread::sleep(Duration::from_secs(1))

# asbtract
yep, differently from the previous mqtt client I made, which had its own event loop waiting for the `strem of messages` this one I thought that just by `starting with callbacks` would create an event loop for itself, but I was wrokg.

--- 

# solution
## to solve the issue I first changed the `start_client` method to:
```rust
pub fn init_client(&mut self) {
        let lwt = mqtt::Message::new("maximo/lwt", "[LWT] Gateway subscriber lost connection", QOS_1);
        let conn_opts = mqtt::ConnectOptionsBuilder::with_mqtt_version(MQTT_VERSION_5)
            .clean_start(false)
            .properties(mqtt::properties![mqtt::PropertyCode::SessionExpiryInterval => 3600])
            .will_message(lwt)
            .finalize();
        self.client.connect_with_callbacks(conn_opts, on_connect_success, on_connect_failure);
        loop{
            sleep(Duration::from_secs(1))
        }
    }
```

## and the `spawn_mqtt_client` function to:

```rust
pub fn spawn_mqtt_client(window: Window, client_id: i32, url: String){
    tauri::async_runtime::spawn(async move{
        let mut client = &mut MqttClient::new(url.clone(), client_id, window);
        client.setup_client();
        client.init_client();
    });
}
```

---

# consideration

about the `Window` and `tauri` u saw in the previous examples, tauri is a desktop app framework which is very useful since it uses the native webview of the OS which makes the app faster and smaller in bundle size since u don't need to vendorize it. it generates the installer for all major OS, and has its own Updater for releases. Window is just a way of accessing tauri's internal `emit` and `listen` functions to have front-back communication in real time(such as an mini mqtt pubsub between the two sides)

tauri: [https://tauri.app/](url)

**Comment by fpagliughi** [comment]
@ArthurFleischman I'm aware of Tauri, but have never tried it out. I'm not much of a GUI guy. Mostly I was just trying to get a minimal example working. 
But what are you building with Tauri and MQTT?

**Comment by ArthurFleischman** [comment]
> @ArthurFleischman I'm aware of Tauri, but have never tried it out. I'm not much of a GUI guy. I pulled it out because, I was just trying to get a minimal example working. But what are you building with Tauri and MQTT?

Mqtt dashboard for IOT in industry complex.
